rust/chains/abacus-ethereum/src/retrying.rs

```diff
@@ -154,6 +133,102 @@ impl JsonRpcClient for RetryingProvider<Http> {
     }
 }
 
+/// Error type for the RetryingProvider
+#[derive(Error, Debug)]
+pub enum RetryingProviderError<P>
+where
+    P: JsonRpcClient,
+{
+    /// An internal error in the JSON RPC Client which we did not want to retry
+    /// on.
+    #[error(transparent)]
+    JsonRpcClientError(P::Error),
+    /// Hit max requests
+    #[error("Hit max requests")]
+    MaxRequests(P::Error),
+}
+
+impl<P> From<RetryingProviderError<P>> for ProviderError
+where
+    P: JsonRpcClient + 'static,
+    <P as JsonRpcClient>::Error: Send + Sync,
+{
+    fn from(src: RetryingProviderError<P>) -> Self {
+        ProviderError::JsonRpcClientError(Box::new(src))
+    }
+}
+
+#[async_trait]
+impl JsonRpcClient for RetryingProvider<Http> {
+    type Error = RetryingProviderError<Http>;
+
+    #[instrument(level = "error", skip_all, fields(method = %method))]
+    async fn request<T, R>(&self, method: &str, params: T) -> Result<R, Self::Error>
+    where
+        T: Debug + Serialize + Send + Sync,
+        R: DeserializeOwned,
+    {
+        self.request_with_retry::<T, R>(method, params, |res, attempt, next_backoff_ms| match res {
+            Ok(res) => HandleMethod::Accept(res),
+            Err(HttpClientError::ReqwestError(e)) => {
```

Comment:
Weird how there are so few `HttpClientError` variants but then the QuorumProvider has so many variants with its `ProviderError`. Wonder why that is

Reply:
Sadly there seems to be a lack of docs on the variants so my guess as to what they are is purely their name which is not always clear enough I would say. I am not even sure if they actually use all of them or if some of them are future proofing. I could take a look but I figure that either way we should handle them as we think makes sense because even future proofed ones are presumably going to get hooked up at some point.